### PR TITLE
Add inference for typing.NamedTuple

### PR DESCRIPTION
**Pull request by euresti**
### Fixes / new features
- Adds support so typing.NamedTuple(...) actually infers something useful.

### CODE REVIEW

File: astroid/brain/brain_typing.py [hunk 1]
```diff
@@ -0,0 +1,43 @@
...
+        raise UseInferenceDefault
+
+    typename = node.args[0].as_string()
+    names = []
+    try:
+        for elt in node.args[1].elts:
+            if len(elt.elts) != 2:
+                raise UseInferenceDefault
+            names.append(elt.elts[0].as_string())
+    except AttributeError:
```
Comment (PCManticore) [inline astroid/brain/brain_typing.py]: Why would this block raise AttributeError?

Comment (euresti) [inline astroid/brain/brain_typing.py]: Well I'm not checking whether `node.args[1]` actually has an `elts` Or whether that elt has more elts in `elt.elts[0]`

File: astroid/brain/brain_typing.py [hunk 2]
```diff
@@ -0,0 +1,43 @@
...
+
+from astroid import (
+    MANAGER, UseInferenceDefault, extract_node, inference_tip,
+    nodes)
+
+def infer_typing_namedtuple(node, context=None):
+    """Infer a typing.NamedTuple(...) call."""
+    # This is essentially a namedtuple with different arguments
+    # so we extract the args and infer a named tuple.
+    func = next(node.func.infer())
```
Comment (euresti) [inline astroid/brain/brain_typing.py]: I'll fix this.

File: astroid/tests/unittest_brain.py [hunk 3]
```diff
@@ -548,6 +548,74 @@ def test_sys_streams(self):
             self.assertIsInstance(raw, astroid.Instance)
             self.assertEqual(raw.name, 'FileIO')
 
+class TypingBrain(unittest.TestCase):
```
Comment (rogalski) [inline astroid/tests/unittest_brain.py]: Decorate it with `@test_utils.require_version('3.6')`

File: astroid/brain/brain_typing.py [hunk 4]
```diff
@@ -0,0 +1,47 @@
...
+    except InferenceError:
+        raise UseInferenceDefault()
+
+    if func.qname() != 'typing.NamedTuple':
+        raise UseInferenceDefault
+
+    if len(node.args) != 2:
+        raise UseInferenceDefault
+
+    typename = node.args[0].as_string()
```
Comment (rogalski) [inline astroid/brain/brain_typing.py]: Maybe moving it just above actual usage is a good idea?

File: astroid/brain/brain_typing.py (hunk 1 above)
Comment (rogalski) [inline astroid/brain/brain_typing.py]: I'd limit catching AttributeErrpr just to `node.args[1].elts`.

Can you clarify second statement in comment above?

**Comment by rogalski** [comment]
@euresti are you willing to make tests all tests pass and finish up this change? It would be nice to have it before releasing Pylint 1.7.0 (firs release wih Python 3.6 support).

**Comment by euresti** [comment]
Sorry for the delay I'll work on this today or tomorrow.

File: astroid/brain/brain_typing.py (hunk 1 above)
Comment (euresti) [inline astroid/brain/brain_typing.py]: `node.args[1].elts` can raise an AttributeError if node.args[1] is not a Tuple.
`elt.elts` can also raise an AttributeError if node.args[1] is a Tuple but doesn't have a Tuple inside it.

I can add a bunch of isinstance checks if you like.

File: astroid/tests/unittest_brain.py (hunk 3 above)
Comment (euresti) [inline astroid/tests/unittest_brain.py]: Technically typing is available in python2 also but in pypi.  Though for tests maybe I should limit it to 3.6

File: astroid/brain/brain_typing.py (hunk 4 above)
Comment (euresti) [inline astroid/brain/brain_typing.py]: Will do

**Comment by euresti** [comment]
Ok hope that's good enough.

**Comment by PCManticore** [comment]
Thanks so much @euresti for this! Thank you @rogalski for helping with the review.

**Pull request merged**